src/REVLoans.sol

```diff
@@ -354,46 +348,38 @@ contract REVLoans is ERC721, ERC2771Context, IREVLoans {
     /// @param collateralToReturn The amount of collateral to return being returned from the loan.
     /// @param beneficiary The address receiving the returned collateral and any tokens resulting from paying fees.
     /// @param allowance An allowance to faciliate permit2 interactions.
-    /// @return loan The loan after it's been paid off.
+    /// @return paidOffLoanId The ID of the loan after it's been paid off.
+    /// @return paidOffloan The loan after it's been paid off.
     function payOff(
         uint256 loanId,
         uint256 amount,
         uint256 collateralToReturn,
         address payable beneficiary,
-        JBSingleAllowance memory allowance
+        JBSingleAllowance calldata allowance
     )
         external
         payable
         override
-        returns (REVLoan memory)
+        nonReentrant
+        returns (uint256, REVLoan memory)
     {
         // Make sure only the loan's owner can manage it.
         if (_ownerOf(loanId) != _msgSender()) revert UNAUTHORIZED();
 
         // Keep a reference to the fee being iterated on.
         REVLoan storage loan = _loanOf[loanId];
 
-        // Keep a reference to the time since the loan was created.
-        uint256 timeSinceLoanCreated = block.timestamp - loan.createdAt;
-
-        // Keep a reference to the fee that'll be taken.
-        uint256 sourceFeeAmount;
-
-        // If the loan period has passed the prepaid time frame, take a fee.
-        if (timeSinceLoanCreated > loan.prepaidDuration) {
-            // If the loan period has passed the liqidation time frame, do not allow loan management.
-            if (timeSinceLoanCreated > LOAN_LIQUIDATION_DURATION) revert LOAN_EXPIRED();
-
-            // Calculate the prepaid fee for the amount being paid back.
-            uint256 prepaidAmount = JBFees.feeAmountFrom({amount: amount, feePercent: loan.prepaidFeePercent});
-
-            // Calculate the fee as a linear proportion given the amount of time that has passed.
-            sourceFeeAmount = mulDiv(amount, timeSinceLoanCreated, LOAN_LIQUIDATION_DURATION) - prepaidAmount;
-        }
+        if (collateralToReturn > loan.collateral) revert COLLATERAL_EXCEEDS_LOAN();
 
         // Accept the funds that'll be used to pay off loans.
         amount = _acceptFundsFor({token: loan.source.token, amount: amount, allowance: allowance});
 
+        // Keep a reference to the fee that'll be taken.
+        uint256 sourceFeeAmount = _determineSourceFeeAmount(loan, amount);
+
+        // Make sure the amount being paid off is less than the loan's amount.
+        if (amount - sourceFeeAmount > loan.amount) revert AMOUNT_EXCEEDS_LOAN();
+
         // If the amount being paid is greater than the loan's amount, return extra to the payer.
         if (amount > loan.amount + sourceFeeAmount) {
             _transferFrom({
```

Comment:
Don't these two contradict each other?

Reply:
nice. i left the transferFrom, and reset `amount -= (sourceFeeAmount + loan.amount)`